pkg/graveler/graveler.go

```diff
@@ -2528,6 +2530,124 @@ func (g *Graveler) Merge(ctx context.Context, repository *RepositoryRecord, dest
+	// or some other branch changing operation. If commit is in-progress, then staging area wasn't empty after we checked so not retrying is ok.
+	// If another commit/merge succeeded, then the user should decide whether to retry the merge.
+	err := g.retryBranchUpdate(ctx, repository, destination, func(branch *Branch) (*Branch, error) {
+		empty, err := g.isSealedEmpty(ctx, repository, branch)
+		if err != nil {
+			return nil, fmt.Errorf("check if staging empty: %w", err)
+		}
+		if !empty {
+			return nil, fmt.Errorf("%s: %w", destination, ErrDirtyBranch)
+		}
+		toCommit, err := g.dereferenceCommit(ctx, repository, Ref(destination))
+		if err != nil {
+			return nil, err
+		}
+
+		// Source is not a commit - merge base is destination root
+		itr, err := g.Log(ctx, repository, toCommit.CommitID, false)
+		if err != nil {
+			return nil, err
+		}
+		root := toCommit
+		for itr.Next() {
+			root = itr.Value()
+		}
+
+		g.log(ctx).WithFields(logging.Fields{
+			"repository":             repository.RepositoryID,
+			"source":                 source,
+			"destination":            destination,
+			"source_meta_range":      source,
+			"destination_meta_range": toCommit.MetaRangeID,
+			"base_meta_range":        root,
+		}).Trace("Merge")
+
+		metaRangeID, err := g.CommittedManager.MergeFromMetaRange(ctx, storageNamespace, toCommit.MetaRangeID, root.MetaRangeID, source, MergeStrategySrc)
+		if err != nil {
+			if !errors.Is(err, ErrUserVisible) {
+				err = fmt.Errorf("merge in CommitManager: %w", err)
+			}
+			return nil, err
+		}
+		commit = NewCommit()
+		commit.Committer = commitParams.Committer
+		commit.Message = commitParams.Message
+		commit.MetaRangeID = metaRangeID
+		commit.Parents = []CommitID{toCommit.CommitID}
+		commit.Generation = toCommit.Generation + 1
+		commit.Metadata = commitParams.Metadata
+		commit.Metadata[MergeStrategyMetadataKey] = MergeStrategySrcWinsStr
+		preRunID = g.hooks.NewRunID()
+		err = g.hooks.PreMergeHook(ctx, HookRecord{
+			EventType:        EventTypePreMerge,
+			RunID:            preRunID,
+			RepositoryID:     repository.RepositoryID,
+			StorageNamespace: storageNamespace,
+			BranchID:         destination,
+			SourceRef:        toCommit.CommitID.Ref(), // Will allow running pre-merge hooks of source ref
+			Commit:           commit,
+		})
+		if err != nil {
+			return nil, &HookAbortError{
+				EventType: EventTypePreMerge,
+				RunID:     preRunID,
+				Err:       err,
+			}
+		}
+		commitID, err = g.RefManager.AddCommit(ctx, repository, commit)
+		if err != nil {
+			return nil, fmt.Errorf("add commit: %w", err)
+		}
+
+		tokensToDrop = branch.SealedTokens
+		branch.SealedTokens = []StagingToken{}
+		branch.CommitID = commitID
+		return branch, nil
+	})
+	if err != nil {
+		return "", fmt.Errorf("update branch %s: %w", destination, err)
+	}
+
+	g.dropTokens(ctx, tokensToDrop...)
```

Comment:
merge is on committed data - we are dropping stage tokens at the end as we committed these changes.

Reply:
**This code code was not modified**, if you think there's an issue here we should address it in a separate PR